bestie/server/testresult.go

```diff
@@ -0,0 +1,220 @@
+type bazelStream struct {
+	buildId       string
+	invocationId  string
+	testName      string
+	run           string
+	invocationSha string // derived
+}
+
+// Derive a unique invocation SHA value.
+func deriveInvocationSha(items []string) string {
+	// The SHA value allows combining one or more items
+	// (e.g. invocationId, buildId, run) to uniquely identify
+	// a bazel test stream without impacting the database schema.
+	hash := sha256.Sum256([]byte(strings.Join(items[:], ".")))
+	return hex.EncodeToString(hash[:])
+}
+
+// Store fields that help identify this Bazel stream.
+func identifyStream(bazelBuildEvent bes.BuildEvent, streamId *build.StreamId) bazelStream {
+	// Extract the stream identifier fields of interest.
+	stream := bazelStream{
+		buildId:      streamId.GetBuildId(),
+		invocationId: streamId.GetInvocationId(),
+		run:          strconv.Itoa(int(bazelBuildEvent.GetId().GetTestResult().GetRun())),
+		testName:     bazelBuildEvent.GetId().GetTestResult().GetLabel(),
+	}
+	// Calculate a SHA256 hash using the following fields to uniquely identify this stream.
+	stream.invocationSha = deriveInvocationSha([]string{stream.invocationId, stream.buildId, stream.run})
+	return stream
+}
+
+// Handle metrics extraction from the TestResult event.
+func handleTestResultEvent(bazelBuildEvent bes.BuildEvent, streamId *build.StreamId) error {
+	stream := identifyStream(bazelBuildEvent, streamId)
+	m := bazelBuildEvent.GetTestResult()
+	if m == nil {
+		return fmt.Errorf("Error extracting TestResult data from event message")
+	}
+
+	fmt.Printf("TestResult for %s: %s\n", stream.testName, m.GetStatus())
+	fmt.Printf("  run: %s\n", stream.run)
+	fmt.Printf("  buildId: %s\n", stream.buildId)
+	fmt.Printf("  invocationId: %s\n", stream.invocationId)
+	fmt.Printf("  invocationSha: %s\n\n", stream.invocationSha)
+
+	outputFiles := m.GetTestActionOutput()
+	var ofname, ofuri string
+	found := false
+	for _, of := range outputFiles {
+		ofname = of.GetName()
+		ofuri = of.GetUri()
+		if strings.HasSuffix(ofname, "outputs.zip") {
+			found = true
+			break
+		}
+	}
+	if !found {
+		return nil
+	}
+
+	// Strip off any file:// prefix from the URI to access the local file system path.
+	filePrefix := "file://"
+	if strings.HasPrefix(ofuri, filePrefix) {
+		ofuri = ofuri[len(filePrefix):]
+	}
+
+	// Process test metrics output file(s).
+	// Each output file contains a single (potentially large) protobuf message.
+	if err := extractZippedFiles(stream, ofuri); err != nil {
+		return fmt.Errorf("Error processing %s file: %s", ofname, err)
+	}
+
+	return nil
+}
+
+// Look for and return a []byte slice with its contents.
+func extractZippedFiles(stream bazelStream, zipFile string) error {
+	metricsFileRE := regexp.MustCompile(`(^test_metrics.pb$|^test_metrics[_-]+[[:word:]-]*[[:alnum:]]\.pb$)`)
+	summaryFileRE := regexp.MustCompile(`(^test_summary.pb$|^test_summary[_-]+[[:word:]-]*[[:alnum:]]\.pb$)`)
```

Comment:
What is a summary file - how is it different than a metrics file?

Reply:
The intent is to put actual test metrics in a "test_metrics*.pb" file and the individual test case pass/fail summaries in a "test_summary*.pb" file. They are formatted the same, but the pass/fail summary "metrics" use a fixed metric name of 'testresult' and a value of 1.0 or 0.0 to indicate pass vs. fail, staying consistent with both the protobuf message definition and the BigQuery table schema, since we want these all to land in the same BigQuery test metrics table. With the regex change mentioned in the previous comment, the actual file names don't need to be as rigid.